design-proposals/instancetype.kubevirt.io/support-declarative-vm-management.md

```diff
@@ -0,0 +1,294 @@
+
+The same will also be possible when moving between generations of an instance type if the new generation introduces additional supported resources such as vCPUs exposed as sockets or memory.
+
+## API Examples
+
+For a simple VirtualMachine using a VirtualMachineClusterInstancetype named `foo` the following details should end up in status:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When moving between instance types users will now only need to modify the name (and kind if originally provided) with the VirtualMachine controller then attempting to reconcile the spec with the status of the VirtualMachine:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+A user patches `spec.instancetype.name` with `fooNew`, `foo` is still referenced by `status.instancetypeRef.name`:
+
+```yaml
+spec:
+  instancetype:
+    name: fooNew
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+Eventually the VirtualMachine controller aligns `status.instancetypeRef.name` with `spec.instancetype.name`:
+
+```yaml
+spec:
+  instancetype:
+    name: fooNew
+[..]
+status
+  instancetypeRef:
+    name: fooNew
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When `kind` and/or `revisionName` are already populated in existing VirtualMachines these values are mirrored into status by the controller:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    revisionName: bar
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+This will also happen with `controllerRevisionRef` in the spec:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When using `inferFromVolume` with a new VirtualMachine the Matcher within the spec is no longer mutated with all runtime derived details now stashed in status:
+
+```yaml
+spec:
+  instancetype:
+    inferFromVolume: foo
+status
+  instancetypeRef:
+    name: derivedName
+    kind: derivedKind
+    controllerRevisionRef:
+      name: derivedRevisionName
+
+```
+
+If a snapshot is taken of this VirtualMachine however the values from status will be written back into the spec with `inferFromVolume` now removed:
```

Comment:
It will lead to slightly different behavior of a restored VM. On the original VM the inference can be retriggered, on the restored VM the instancetype/preference will be fixed. Is there a way to keep `inferFromVolume` and only restore the status?

Reply:
We could always move `name`, `kind`, `RevisionName` and `controllerRevisionRef` into `status` when `inferFromVolume` is provided? That's simple enough and would allow the restored VM to retain the same `inferFromVolume` behaviour.

Reply:
Actually that doesn't work, we can't change the spec after submission at all even in this case. I guess if the user wants `inferFromVolume` to work again this case they will need to clear the other fields in the spec and then use the `refresh` subresource API etc to infer things once again.